In [1]:
import numpy as np
import pandas as pd

In [17]:
users = pd.read_csv("./data/users_preprocessed.csv")
users.nunique()

user_id             68092
age                   364
location_city        1560
location_state        399
location_country      199
dtype: int64

# feature 카테고리 수 줄이기
- features
    - 'user_id', 
    - 'isbn'
    - 'age' : 범주화해서 사용 (6)
    - 'location_city' : x (1560)
    - 'location_state' : x (399)
    - 'location_country' : o (199)
    - 'year_of_publication' : 범주화해서 사용 (5)
    - 'new_language' : 'isenglish' (2)
    - 'category_high' : 'isfiction' (2)
    - 'book_author' : 'book_author_over100' (34)
    - 'publisher' : x
    

In [8]:
books = pd.read_csv('./data/books.csv')
books.nunique()

isbn                   149570
book_title             135436
book_author             62059
year_of_publication        95
publisher               11571
img_url                149570
language                   26
category                 4292
summary                 79872
img_path               149570
dtype: int64

In [10]:
books.shape

(149570, 10)

In [14]:
books_is = pd.read_csv('./data/isen_isfic_books.csv')

In [15]:
books_is.shape

(149570, 17)

In [16]:
books_is.nunique()

Unnamed: 0             149570
isbn                   149570
book_title             135436
year_of_publication        95
publisher               11571
img_url                149570
language                   26
summary                 79872
img_path               149570
category_high             749
book_author             61828
book_author_over10       2105
category                 4231
new_language               41
remove_country_code    148470
isfiction                   2
isenglish                   2
dtype: int64

In [4]:
books_merged = pd.read_csv('./data/books_merged copy.csv')

In [11]:
books_merged.shape

(149570, 18)

In [6]:
books_merged.nunique()

isbn                   149570
book_title             135436
year_of_publication        95
publisher               11571
img_url                149570
language                   26
summary                 79872
img_path               149570
category_high             412
book_author             61828
category                 4263
new_language               41
remove_country_code    148470
book_author_over3       10755
book_author_over5        5419
book_author_over10       2105
book_author_over50        172
book_author_over100        34
dtype: int64

In [13]:
books_merged2 = pd.read_csv('./data/books_merged_2.csv')
books_merged2.nunique()

isbn                   149570
book_title             135436
year_of_publication        95
publisher                1523
img_url                149570
language                   26
summary                 79872
img_path               149570
category_high             412
book_author             61828
book_author_over10       2105
category                 4263
new_language               41
remove_country_code    148470
dtype: int64

In [22]:
new_books = books_merged2.drop(['book_title','img_url','summary'], axis=1).merge(books_is[['isbn','isfiction','isenglish']], how='left', on='isbn').merge(books_merged[['isbn','book_author_over3','book_author_over5','book_author_over50','book_author_over100']], how='left',on='isbn')

In [24]:
new_books.nunique()

isbn                   149570
year_of_publication        95
publisher                1523
language                   26
img_path               149570
category_high             412
book_author             61828
book_author_over10       2105
category                 4263
new_language               41
remove_country_code    148470
isfiction                   2
isenglish                   2
book_author_over3       10755
book_author_over5        5419
book_author_over50        172
book_author_over100        34
dtype: int64

In [50]:
pd.set_option('display.max_rows', 500)
def year_to_map(x):
    if x < 1970:
        return 0
    elif x < 1980:
        return 1
    elif x < 1990:
        return 2
    elif x < 2000:
        return 3
    else:
        return 4
new_books['new_year'] = new_books['year_of_publication'].map(year_to_map)

In [51]:
new_books['new_year'].value_counts()

3    72358
4    46301
2    23809
1     5636
0     1466
Name: new_year, dtype: int64

In [67]:
pd.set_option('display.max_rows', 1600)
tmp = new_books['publisher'].value_counts()
new_books['publisher_small'] = new_books['publisher'].copy()
for id in tmp.index:
    if tmp[id] < 1000:
        new_books.loc[new_books['publisher']==id, 'publisher_small'] = 'others'

In [70]:
new_books.nunique()

isbn                   149570
year_of_publication        95
publisher                1523
language                   26
img_path               149570
category_high             412
book_author             61828
book_author_over10       2105
category                 4263
new_language               41
remove_country_code    148470
isfiction                   2
isenglish                   2
book_author_over3       10755
book_author_over5        5419
book_author_over50        172
book_author_over100        34
new_year                    5
publisher_small            26
dtype: int64

In [71]:
new_books.to_csv("./data/books_final.csv", index=False)

In [2]:
tmp = pd.read_csv('./data/books_embedding_2.csv')

In [4]:
books_new = pd.read_csv('./data/books_final.csv')

In [8]:
tmp.filter(regex=r'(^\d+$)|(isbn)')

isbn         0         1         2         3         4  \
0       0002005018  0.503390  0.231714  0.115370 -0.329476  0.266069   
1       0060973129 -0.689147  0.700956  1.187890  0.556178  1.047582   
2       0374157065 -0.567941 -0.759501 -0.035331  0.337106 -0.449440   
3       0399135782  1.013494  0.018726  0.398491  0.193576 -0.090819   
4       0425176428  0.886252 -0.758539  0.158425 -0.916154 -0.395121   
...            ...       ...       ...       ...       ...       ...   
149565  067161746X  0.793568 -0.433235  0.336982  0.172376  0.645219   
149566  0767907566 -0.504454  0.322140  0.080038 -0.428775  0.231001   
149567  0884159221  0.302760  0.249366  0.411577 -0.741776 -0.362098   
149568  0912333022  0.484288 -0.404645 -1.898071 -0.826890 -0.576069   
149569  1569661057 -0.547346 -0.119047  0.048054  0.457854 -0.068116   

               5         6         7         8  ...       502       503  \
0       0.508641 -0.019863  0.510959 -0.317959  ...  0.781929  0.222393   
1       0.735019 -0.491351 -0.441994  0.244543  ...  0.215752  0.093943   
2      -0.044827  0.211133  0.298013  0.229991  ...  0.647762  0.017774   
3      -0.147283 -1.139013 -0.507875  0.461223  ...  1.110681 -0.826044   
4      -1.181997  0.650359 -0.670347  0.487777  ... -0.153706  0.628128   
...          ...       ...       ...       ...  ...       ...       ...   
149565 -0.632824  0.053440 -0.429813  0.373318  ... -0.032002  0.180925   
149566  0.444427  0.115094 -1.053576 -0.202702  ... -0.157287 -0.482961   
149567 -0.633594  0.012307 -0.583492 -0.085389  ... -0.186697 -0.001326   
149568  0.368019 -0.815649 -0.759489 -0.129539  ...  1.398664  0.920493   
149569  0.039923 -0.011214 -0.136012 -0.404763  ...  0.768632  0.013970   

             504       505       506       507       508       509       510  \
0       0.487293  0.068784  0.037725 -1.268462 -0.750312  0.107709 -0.462396   
1       0.151725 -0.537121 -0.453638 -0.858057 -0.290574 -0.710232 -0.775910   
2      -0.340387 -0.186091 -0.065213  0.268273 -1.180975  0.385781  0.077389   
3      -0.892639  0.164168 -0.205375 -0.245697 -0.806942  1.030093  0.396518   
4       0.500611  0.919893 -1.029661 -1.536474 -1.166991 -0.001767  0.096962   
...          ...       ...       ...       ...       ...       ...       ...   
149565 -0.716202  0.018995 -0.944208  0.684940 -0.474849  0.102360 -0.878273   
149566 -0.342532 -0.147599 -0.832974 -0.287492 -0.298440 -0.493880 -1.217481   
149567 -0.347721 -0.063748 -0.538890  0.096727  0.201310 -0.175871  0.166480   
149568 -0.433893 -0.457107  0.263443 -1.829711  1.697475 -0.153881  0.443192   
149569  0.274785 -0.002244 -0.214558 -0.686849 -0.300046  0.028379 -0.531429   

             511  
0       0.383121  
1      -0.211170  
2       0.683517  
3       0.037680  
4       0.329258  
...          ...  
149565 -0.836812  
149566 -0.157467  
149567 -0.597726  
149568 -1.214950  
149569  0.354587  

[149570 rows x 513 columns]

In [9]:
books_new.merge(tmp.filter(regex=r'(^\d+$)|(isbn)'),'left','isbn').to_csv('./data/books_final_embedding.csv', index=False)

In [3]:
tmp.head()

isbn                                         book_title  \
0  0002005018                                       Clara Callan   
1  0060973129                               Decision in Normandy   
2  0374157065  Flu: The Story of the Great Influenza Pandemic...   
3  0399135782                             The Kitchen God's Wife   
4  0425176428  What If?: The World's Foremost Military Histor...   

            book_author  year_of_publication  \
0  Richard Bruce Wright               2001.0   
1          Carlo D'Este               1991.0   
2      Gina Bari Kolata               1999.0   
3               Amy Tan               1991.0   
4         Robert Cowley               2000.0   

                                             img_url language   category  \
0  http://images.amazon.com/images/P/0002005018.0...       en  actresses   
1  http://images.amazon.com/images/P/0060973129.0...       en  1940-1949   
2  http://images.amazon.com/images/P/0374157065.0...       en    medical   
3  http://images.amazon.com/images/P/0399135782.0...       en    fiction   
4  http://images.amazon.com/images/P/0425176428.0...       en    history   

                                             summary  \
0  In a small town in Canada, Clara Callan reluct...   
1  Here, for the first time in paperback, is an o...   
2  Describes the great flu epidemic of 1918, an o...   
3  A Chinese immigrant who is convinced she is dy...   
4  Essays by respected military historians, inclu...   

                            img_path new_language  ...       505       506  \
0  images/0002005018.01.THUMBZZZ.jpg           en  ...  0.068784  0.037725   
1  images/0060973129.01.THUMBZZZ.jpg           en  ... -0.537121 -0.453638   
2  images/0374157065.01.THUMBZZZ.jpg           en  ... -0.186091 -0.065213   
3  images/0399135782.01.THUMBZZZ.jpg           en  ...  0.164168 -0.205375   
4  images/0425176428.01.THUMBZZZ.jpg           en  ...  0.919893 -1.029661   

        507       508       509       510       511                 publisher  \
0 -1.268462 -0.750312  0.107709 -0.462396  0.383121                   Collins   
1 -0.858057 -0.290574 -0.710232 -0.775910 -0.211170                 Perennial   
2  0.268273 -1.180975  0.385781  0.077389  0.683517      Farrar Straus Giroux   
3 -0.245697 -0.806942  1.030093  0.396518  0.037680          Putnam Pub Group   
4 -1.536474 -1.166991 -0.001767  0.096962  0.329258  Berkley Publishing Group   

   book_author_over10  category_high  
0              others      actresses  
1              others         others  
2              others        medical  
3             Amy Tan        fiction  
4              others        history  

[5 rows x 526 columns]

In [11]:
sum([64,64, 4,4,4,4,4,4,4,4,4,4,4]) + 512

684